libraries/Filter/HarmonicNotchFilter.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "HarmonicNotchFilter.h"
+
+// table of user settable parameters
+const AP_Param::GroupInfo HarmonicNotchFilterParams::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Enable
+    // @Description: Enable dynamic notch filter
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLE", 1, HarmonicNotchFilterParams, _enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: BASEHZ
+    // @DisplayName: Base Frequency
+    // @Description: Notch base center frequency in Hz
+    // @Range: 10 400
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("BASEHZ", 2, HarmonicNotchFilterParams, _center_freq_hz, 80),
+
+    // @Param: BW
+    // @DisplayName: Bandwidth
+    // @Description: Dynamic notch bandwidth in Hz
+    // @Range: 5 100
+    // @Units: Hz
+    // @User: Advanced
+    AP_GROUPINFO("BW", 3, HarmonicNotchFilterParams, _bandwidth_hz, 20),
+
+    // @Param: ATT
+    // @DisplayName: Attenuation
+    // @Description: Dynamic notch attenuation in dB
+    // @Range: 5 30
+    // @Units: dB
+    // @User: Advanced
+    AP_GROUPINFO("ATT", 4, HarmonicNotchFilterParams, _attenuation_dB, 15),
+
+    // @Param: HMNCS
+    // @DisplayName: Harmonics
+    // @Description: Number of harmonic frequencies to add to the filter. This option takes effect on the next reboot.
+    // @Range: 0 2
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("HMNCS", 5, HarmonicNotchFilterParams, _harmonics, 1),
+
+    AP_GROUPEND
+};
+
+
+/*
+  initialise filter
+ */
+template <class T>
+void HarmonicNotchFilter<T>::init(float _sample_freq_hz, float _center_freq_hz, float _bandwidth_hz, float _attenuation_dB)
+{
+    if (filters == nullptr) {
+        return;
+    }
+
+    sample_freq_hz = _sample_freq_hz;
+    bandwidth_hz = _bandwidth_hz;
+    attenuation_dB = _attenuation_dB;
+
+    // adjust the center frequency to be in the allowable range
+    _center_freq_hz = constrain_float(_center_freq_hz, _bandwidth_hz / 1.98f, sample_freq_hz * 0.48f);
+
+    float octaves = log2f(_center_freq_hz  / (_center_freq_hz - bandwidth_hz/2.0f)) * 2.0f;
+    float A = powf(10, -attenuation_dB/40.0f);
+    float Q = sqrtf(powf(2, octaves)) / (powf(2,octaves) - 1.0f);
+
```

Comment:
We should not be reusing this code. We could make it a seperate function in the Notch filter and return those values so that way the same code is used in all three places.

Reply:
Normally I am keen not to bounce between virtuals, but in this case it makes it cleaner so have fixed that way.